x-pack/plugins/ingest_manager/server/plugin.ts

```diff
@@ -152,6 +157,35 @@ export class IngestManagerPlugin
   }
 
   public async setup(core: CoreSetup, deps: IngestManagerSetupDeps) {
+    const maxConcurrentRequests = 1;
+    let concurrentRequests = 0;
+    const isAgentRequest = (request: KibanaRequest) => {
+      const tags = request.route.options.tags;
+      return tags.includes(AGENT_ROUTE_TAG);
+    };
+    core.http.registerOnPreAuth(
+      (request: KibanaRequest, response: LifecycleResponseFactory, toolkit: OnPreAuthToolkit) => {
+        if (!isAgentRequest(request)) {
+          return toolkit.next();
+        }
+
+        if (concurrentRequests >= maxConcurrentRequests) {
+          return response.customError({ body: 'Too Many Agents', statusCode: 429 });
+        }
+
+        concurrentRequests += 1;
+        return toolkit.next();
+      }
+    );
+    core.http.registerOnPreResponse(
+      (request: KibanaRequest, preResponse: OnPreResponseInfo, toolkit: OnPreResponseToolkit) => {
+        if (isAgentRequest(request) && preResponse.statusCode !== 429) {
```

Comment:
This assumes that only the pre-auth handler could cause a response with status code 429, which isn't absolutely certain. It's possible that one of the HTTP route handlers is replying with a 429 itself, or propagating a 429 from Elasticsearch, which would mess up the counter. 

The `request` object is different from the pre-auth to the pre-response, so I wasn't able to use a `Set` or `WeakSet` to track whether or not this was a 429 that we returned from within the pre-auth handler... Any other ideas?

Reply:
Can the preAuth handler make a value accessible to the preResponse handler?

We're [moving](https://github.com/elastic/kibana/pull/69966) [towards](https://github.com/elastic/kibana/pull/67278) using errors like `FleetTooManyAgentsError` so if there's any way to make `new FleetTooManyAgentsError('Too Many Agents')` available in the response handler it could do `typeof FleetTooManyAgentsError`

Reply:
Can we wrap `Ingest manager` specific route handlers in a function instead of adding global interceptors? https://github.com/elastic/kibana/pull/70495/files#diff-280f58825bd033ffbe7792f8423f6122R88
I'm also not sure that we want to provide access to response data in interceptors. 

Also, I don't think the current solution would work if the system consists of more than one Kibana instance.

Reply:
something along lines (not tested):
```typescript
class Lock {
  constructor(private readonly maxConnections: number = 1) {}
  private counter = 0;
  increase() {
    this.counter += 1;
  }
  decrease() {
    this.counter += 1;
  }
  canHandle() {
    return this.counter < this.maxConnections;
  }
}
const lock = new Lock();
export const concurrencyLimit: RequestHandlerWrapper = (handler) => {
  return async (context, request, response) => {
    if (!lock.canHandle()) {
      return response.customError({ body: 'Too Many Agents', statusCode: 429 });
    }
    try {
      lock.increase();
      return handler(context, request, response);
    } finally {
      lock.decrease();
    }
  };
};

concurrencyLimit(postAgentEnrollHandler);
concurrencyLimit(postAgentCheckinHandler);
```

Reply:
this doesn't let us reject traffic without validating API Keys as [described here](https://github.com/elastic/kibana/pull/70495#issuecomment-652646604), right>

Reply:
Yeah... I haven't seen the comment. Now I'm surprised that the auth has such a huge penalty. Need to think about alternatives we have besides adding another interceptor.

Reply:
Fleet's usage of Kibana APIs is different than our traditional usage. The Elastic Agent will be using Kibana APIs to enroll themselves and retrieve their configuration. As such, we're potentially dealing with thousands of Elastic Agents hitting these APIs... Whether or not this is "critical" is debatable and largely dependent on what the ingest-management team is seeing during their load-testing, but skipping auth reduces the load on Kibana when this circuit breaker is hit.

@roncohen despite the current implementation being imperfect and potentially misinterpreting the `429`, can we perform load-testing with the circuit breaking being done before authentication and after authentication to determine what type of impact this has on Fleet's scalability?

> Would server session help to reduce the number of requests?

I don't think it will, Kibana will still have to make a query to Elasticsearch to return the server-side session document to authenticate the user.

> Does this problem exist for API keys only?

Any call to an Elasticsearch API will incur some performance penalty. However, there are differences in the caching strategy for API Keys vs username/password. Regardless of the type of credentials, performing any unnecessary work before their circuit breaker has a chance to short-circuit the operation is inefficient.

Reply:
> @roncohen despite the current implementation being imperfect and potentially misinterpreting the 429, can we perform load-testing with the circuit breaking being done before authentication and after authentication to determine what type of impact this has on Fleet's scalability?

+1 I'd love to adjust the concurrency value (what should the value be) and merge this into master ASAP so we can test in cloud as we did with long polling